stan/math/prim/scal/fun/grad_2F1.hpp

```diff
@@ -1,42 +1,109 @@
-    void grad_2F1(T& gradA, T& gradC, T a, T b, T c, T z, T precision = 1e-6) {
+    void grad_2F1(T& g_a1, T& g_b1, const T& a1, const T& a2, const T& b1,
+        const T& z, const T& precision = 1e-10, int max_steps = 1e5) {
+      check_2F1_converges("grad_2F1", a1, a2, b1, z);
+
+      using std::log;
       using std::fabs;
+      using std::exp;
+
+      g_a1 = 0.0;
+      g_b1 = 0.0;
+
+      T log_g_old[4];
+      for (T *q = log_g_old; q != log_g_old + 4; ++q)
+        *q = -1.0 * std::numeric_limits<double>::infinity();
+
+      T log_t_old = 0.0;
+      T log_t_new = 0.0;
 
-      gradA = 0;
-      gradC = 0;
+      T log_z = log(z);
 
-      T gradAold = 0;
-      T gradCold = 0;
+      T p = 0.0;
+
+      double log_t_new_sign = 1.0;
+      double log_t_old_sign = 1.0;
+      double log_g_old_sign[4];
+      for (double *q = log_g_old_sign; q != log_g_old_sign + 4; ++q)
+        *q = 1.0;
 
       int k = 0;
-      T tDak = 1.0 / (a - 1);
+      T term;
+      while (true) {
+        p = (a1 + k) * (a2 + k) / ((b1 + k) * (1 + k));
+        if (p == 0)
+          break;
+
+        log_t_new += log(fabs(p)) + log_z;
+        log_t_new_sign *= sign(p);
 
-      while (fabs(tDak * (a + (k - 1)) ) > precision || k == 0) {
-          const T r = ( (a + k) / (c + k) ) * ( (b + k) / (T)(k + 1) ) * z;
-          tDak = r * tDak * (a + (k - 1)) / (a + k);
+//      g_old[0] = t_new * (g_old[0] / t_old + 1.0 / (a1 + k));
+        term = log_g_old_sign[0] * log_t_old_sign *
+          exp(log_g_old[0] - log_t_old) + 1/(a1 + k);
+        log_g_old[0] = log_t_new + log(fabs(term));
+        log_g_old_sign[0] = sign(term) * log_t_new_sign;
 
-          if (r == 0) break;
+//      g_old[1] = t_new * (g_old[1] / t_old + 1.0 / (a2 + k));
+        term = log_g_old_sign[1] * log_t_old_sign *
+          exp(log_g_old[1] - log_t_old) + 1/(a2 + k);
+        log_g_old[1] = log_t_new + log(fabs(term));
+        log_g_old_sign[1] = sign(term) * log_t_new_sign;
 
-          gradAold = r * gradAold + tDak;
-          gradCold = r * gradCold - tDak * ((a + k) / (c + k));
+//      g_old[3] = t_new * (g_old[3] / t_old - 1.0 / (b1 + k));
+        term = log_g_old_sign[2] * log_t_old_sign *
+          exp(log_g_old[2] - log_t_old) - 1/(b1 + k);
+        log_g_old[2] = log_t_new + log(fabs(term));
+        log_g_old_sign[2] = sign(term) * log_t_new_sign;
 
-          gradA += gradAold;
-          gradC += gradCold;
+        g_a1 += log_g_old_sign[0] * exp(log_g_old[0]);
+        g_b1 += log_g_old_sign[2] * exp(log_g_old[2]);
 
-          ++k;
+        if (exp(log_t_new) <= precision)
+          break;  // implicit abs
 
-          if (k > 200) break;
+        if (k >= max_steps) {
+          domain_error("grad_2F1", "k (internal counter)", max_steps,
+            "exceeded ", " iterations, hypergeometric function gradient "
+            "did not converge.");
         }
+
+        log_t_old = log_t_new;
+        log_t_old_sign = log_t_new_sign;
+        ++k;
```

Comment:
Whenver there's a `++k` at the end of a loop, it means you really wanted a for loop.  Here over `k` rather than setting `k` to zero, then running a `while (true)` loop.  The middle condition can just be `true`, but having a for loop makes the initialization and iteration much clearer to the code reader.